Simplify handling of BL-Touch command timing

KevinOConnor opened the pull request:
The BL-Touch timing optimization in commit 5c8d15bb had to be reverted (as of commit a028aeaf).  Here's another attempt at it, which also tries to simplify the overall handling of command timing.  Unfortunately, I don't have a good way to test this, so consider it a bit risky.

@FanDjango - FYI.

-Kevin

FanDjango commented:
@KevinOConnor 

I was moving in this direction as well to alleviate the complexity further down in the code. Actually even the question of "touch-mode" yes/no can be solved more elegantly by deciding the sequence first and then generating the timing.

Always generate the PWM command signal for 100ms and always reserve at
least 100ms of PWM off time before the next command is sent. Handle
this directly from send_cmd().

I am going to stop duplicate working on this code and stick to debugging only. Waste of time to have two people coding and the restriction to create tiny incremental PR's is a hindrance to my workflow.

I will test this as soon as I can.

FanDjango commented:
Ok, first test is a disapointment (`stow_on_each_sample: True` as that is the stress case):
```
Recv: // probe at -13.000,25.000 is z=2.321250
Recv: // probe at 25.750,25.000 is z=2.360000
Recv: // probe at 64.500,25.000 is z=2.383750
Recv: // probe at 103.250,25.000 is z=2.391250
Recv: // probe at 142.000,25.000 is z=2.408750
Recv: // probe at 142.000,63.750 is z=2.453750
Recv: // probe at 103.250,63.750 is z=2.436250
Recv: // probe at 64.500,63.750 is z=2.431250
Recv: // probe at 25.750,63.750 is z=2.396250
Recv: // probe at -13.000,63.750 is z=2.367500
Recv: // probe at -13.000,102.500 is z=2.390000
Recv: // probe at 25.750,102.500 is z=2.427500
Recv: // Failed to verify BLTouch probe is raised; retrying.
Recv: // probe at 64.500,102.500 is z=2.451250
Recv: // Failed to verify BLTouch probe is raised; retrying.
Recv: // probe at 103.250,102.500 is z=2.452500
Recv: // Failed to verify BLTouch probe is raised; retrying.
Recv: // probe at 142.000,102.500 is z=2.478750
Recv: // probe at 142.000,141.250 is z=2.498750
Recv: // probe at 103.250,141.250 is z=2.475000
```
Previous `bltouch.py` version was/is flawless.

Happy Easter!

KevinOConnor commented:
Okay - I don't know why it would result in a regression and it isn't important enough to worry about.  I'll drop this patch for consideration.

-Kevin

fiferboy commented:
@KevinOConnor can we make sure the "touch mode" patch gets integrated somewhere? It is by far the best way for my probe to work and looks like a simple feature to be included (but @FanDjango dropped his pull request in favour of this one).

FanDjango commented:
Slowly, slowly - I like this new code better (cleaner, no more "None" etc.) than the old and it's a better place to put in the "touch mode" patch. Let's find out what is wrong with it and check it out better. I've scrapped all of my stuff and will use this new code as a basis for further experiments and added functionality.

FanDjango commented:
I am getting very good results with the code from this PR after making some small changes:

All of the problems I was observing were actually the same symptoms as under scrutiny in #2690 - failure to raise the pin in time before the probe goes into error state.

The following observations are slightly interdependant.

1. `MIN_CMD_TIME = 8` instead of 5 significantly reduces the error rate **if I make no other changes**. It seems that the probe then has a better chance of recognizing the "pin_up" command after a trigger. _But some errors remained, it was not a final solution_. Note that in Marlin I also arrived at 150ms (instead of 160ms) as the minimum command recognition time.

2. Adding a **different** non-destructive command (like `reset`) in front of the `self.send_cmd('pin_up', duration=self.pin_move_time)` **reduced the error rate to 0**, even if I reverted the MIN_CMD_TIME back to 5. Coded like so:
```
    def raise_probe(self):
        self.sync_mcu_print_time()
        if not self.pin_up_not_triggered:
            # No way to verify raise attempt - just issue commands
            self.send_cmd('reset')
            self.send_cmd('pin_up', duration=self.pin_move_time)
            return
        for retry in range(3):
            self.send_cmd('reset')
            self.send_cmd('pin_up', duration=self.pin_move_time)
            success = self.verify_state(False)
            if success:
                # The "probe raised" test completed successfully
                break
            if retry >= 2:
                raise homing.EndstopError("BLTouch failed to raise probe")
            msg = "Failed to verify BLTouch probe is raised; retrying."
            self.gcode.respond_info(msg)
            self.sync_mcu_print_time()
            self.send_cmd('reset', duration=RETRY_RESET_TIME)
```
It looks silly, but it works.

My theory on this is that maybe the probe's internal code somehow syncs itself on a continuous PWM and needs different amounts of time based on its internal state to "come back into" the process of recognizing the PWM pulse width if PWM is interrupted. 

Note that simply making the MIN_CMD_TIME for the "pin_up" command **did not** help at all, there was no change in the rate of failures to recognize the "pin_up" command. Also there is a limit as to how high you can go on this, as the upper bound is the ~650ms overall trigger to pin-up time limit before you get an error.

Prompted by this I have looked at the code and am wondering how easy or not it would be to have PWM running the "pin_down" command continuously on the probe downwards movement?

Could you add the silly "reset" and the #2688 commits into this one here? Then the lurkers are invited to play with this and give us some input as to how it performs on **their** systems.

KevinOConnor commented:
> My theory on this is that maybe the probe's internal code somehow syncs itself on a continuous PWM and needs different amounts of time based on its internal state to "come back into" the process of recognizing the PWM pulse width if PWM is interrupted.

The device reliably handles pin_down commands even after a long silence on the signal line.  The pin_up command on the master branch today is sent for the full pin_move_time - which is typically 680ms (34 repeats).  Only a small number of pin_up sequences have a problem - the majority of pin_up sequences are recognized immediately by the device.

Given the above, I'd describe the issue as, "the bltouch v3.1 is buggy - there is some window where it will not recognize a valid pin_up command and once in that state it will not recognize future repeat pin_up commands either".

There are a few things one could try:
1. Change SIGNAL_PERIOD to .023456 - perhaps a different period will avoid the "window".
2. Change the `sync_mcu_time()` to use a different delay - for example use: `self.next_cmd_time = max(self.next_cmd_time, est_time + .123456)` - perhaps a different delay from trigger point will avoid the "window".
3. Try probing in "touch_mode" to see if that doesn't have the "window" at all.  (I've added the probe_with_touch_mode option to this branch.)

It's also worth considering that the final solution may be to "do nothing".  Users of these devices may have to accept that they are buggy and they will have an occasional "reset" sequence mid probe.

-Kevin

FanDjango commented:
> It's also worth considering that the final solution may be to "do nothing". Users of these devices may have to accept that they are buggy and they will have an occasional "reset" sequence mid probe.

Yes. Time to check out everything with the V2.0 etc. as a sanity check. Beginning to hate the V3.1 myself. Only one thing that is even more frustrating: Marlin handles it ok. 

> Try probing in "touch_mode" to see if that doesn't have the "window" at all. (I've added the probe_with_touch_mode option to this branch.)

It's not called SW ( = switch ) mode for nothing. That's the mode that is supposed to make this type of probe more or less a drop in replacement for a real switch.

Now I know why I have been using Touch Mode **and** No Stow/Deploy sequences in between since a year or so. Always hated the "normal" mode. And so I rarely came into contact with these quirks on my machines, only ever when users were suffering and trying to support them.

>  The pin_up command on the master branch today is sent for the full pin_move_time - which is typically 680ms (34 repeats).

Yes, I tried all sorts of durations, didn't make things better. If the probe didn't want to see the signal, it didn't, however long you presented it to the probe.

> Only a small number of pin_up sequences have a problem - the majority of pin_up sequences are recognized immediately by the device.

Yes. I agree. Once in while it just refuses. I needed all that debugging output to see that.

> Given the above, I'd describe the issue as, "the bltouch v3.1 is buggy - there is some window where it will not recognize a valid pin_up command and once in that state it will not recognize future repeat pin_up commands either".

Absolutely. 

The remaining question(s) might be: Firstly, why or how (that's my code there we are talking about) does Marlin avoid ever opening that window and secondly, do we care? - the effort is getting to be stupendous. **Why not advise Touch Mode for the V3.1**? (Expressly allowed by ANTClabs, but just like you I am getting to the point where I am just thoroughly disgusted by all them, not high quality code in that stupid thing).

Bad weather coming, so I will give it a few more shots testing your suggestion #1 and #2 and then call it off. I've got all the needed tools for this now thanks to your efforts.

This PR's version is still looking nice and clean, though. And with the touch mode option, others can test it as well, **please**

FanDjango commented:
> Adding a different non-destructive command (like reset) in front of the self.send_cmd('pin_up', duration=self.pin_move_time) reduced the error rate to 0, even if I reverted the MIN_CMD_TIME back to 5

In case you haven't seen, @lolejar in #2690 already testet and the MIN_CMD_TIME to 8 helped quite a bit, but the second test did not. 

Maybe you should consider giving **all** type of probes more than just 5 cycles, 8 instead. **It did have quite some influence.**

KevinOConnor commented:
FYI, I rebased this branch - the code will no longer limit all commands to only  100ms of transmit time.  Those commands that use pin_move_time will keep transmitting up to that time.  Ideally, this will make this branch less likely to introduce a regression.

As before, though, I don't have any particular attachment to this change.  If it's thought to be useful, great.  If not, then I'm fine with dropping it.

-Kevin

KevinOConnor commented:
> Firstly, why or how (that's my code there we are talking about) does Marlin avoid ever opening that window 

If you want my guess, it's the 100ms delay from endstop signal to first pin_up.  I'd guess that Marlin responds quicker and thus never hits the "window".  For example (wild speculation follows), if the bltouch v3.1 is periodically issuing "internal pin_down" sequences after a trigger event, then a pin_up command that does not come quickly may end up racing with that internal processing.

-Kevin

FanDjango commented:
I like this PR because the code gets less complex and touch mode would be the potential solution for some BLTouch V3.1 problems.

But I would agree totally to avoid all regressions and let this PR and its branch sit there as a test case for all future BLTouch issues to check out viability until a final decision is taken.

Since it encompasses only `bltouch.py` changes, + a doc change, it is easy to rebase once in a while.

Leave it sitting there for a maximum of some months and a decision will become easier.

fiferboy commented:
The new timing is looking good for my seemingly wonky clone probe! I took out the touch mode from the config to try the regular pin drop mode and did a sequence of 25 then 50 accuracy tests without a single misfire.

```
Send: PROBE_ACCURACY SAMPLES=50
Recv: // PROBE_ACCURACY at X:155.000 Y:107.000 Z:8.000 (samples=50 retract=5.000 speed=5.0 lift_speed=10.0)
Recv: // probe at 155.000,107.000 is z=3.425676
Recv: // probe at 155.000,107.000 is z=3.436505
...
Recv: // probe at 155.000,107.000 is z=3.424010
Recv: // probe accuracy results: maximum 3.436505, minimum 3.424010, range 0.012495, average 3.429941, median 3.429841, standard deviation 0.003151
```

I then switched to touch mode (which is just my preference) and it worked just as well. I don't know if there is anything else I should test (more repetitions? revert and retest?) but it looks like a success to me.

pusewicz commented:
I've been giving this branch a spin, loving the touch mode probing (with the original BLTouch).

```
Recv: // probe accuracy results: maximum 2.140000, minimum 2.130000, range 0.010000, average 2.134300, median 2.135000, standard deviation 0.001735
```

KevinOConnor commented:
FYI, I've rebased this branch and added documentation on both `probe_in_touch_mode` and `stow_on_each_sample`.  The code is otherwise unchanged.  It would be great if the documentation could be reviewed.

Unless there is an objection, I'm inclined to commit these changes as the result has been overall positive.

-Kevin

pusewicz commented:
The docs update looks good to me 👍

FanDjango commented:
> Leave it sitting there for a maximum of some months and a decision will become easier.

Pretty close now to be able to say: Check the hardware (when a BLTouch problem occurs).
